replacer.go

```diff
@@ -125,6 +125,16 @@ func (r *Replacer) replace(input, empty string,
 	// iterate the input to find each placeholder
 	var lastWriteCursor int
 	for i := 0; i < len(input); i++ {
+
+		// check for the escape character
+		if i > 0 && input[i-1] == phEscape {
+			if input[i] == phClose || input[i] == phOpen {
+				sb.WriteString(input[lastWriteCursor : i-1])
+				lastWriteCursor = i
+			}
+			continue
+		}
+
```

Comment:
For some reason, the placement of this block isn't sitting well with me. My bug sense is tingling. Maybe new test cases will help validate or dissolve my concerns.

I'm not sure that `continue` is in the right place. Does that skip all backslashes?

Doesn't this immediately write all content (minus backslash) since the `lastWriteCursor`, and then skip all the logic below? I feel like we need to maybe put backslash-skipping logic around existing `WriteString` calls. (Does that make sense?)

I'm just supposing here -- sorry if it turns out to be wrong!

Reply:
This makes a lot of sense and I do believe there is a bug here. Will confirm this with some tests and the fix before considering code placement.

Reply:
Interestingly this works as expected. It doesn't skip all backslashes as `lastWriteCursor` isn't modified and so they are picked up in the next write. That said, I've collapsed the `if` block as the way it is structured is open to confusion.

Moving things closer to the existing `WriteString` calls works well for the check around opening braces but I haven't yet found a way to avoid making the check for the closing brace more complex. Will continue to think about this one.

Reply:
Okay. Does that mean you want to continue working on this PR? Or is it finished do you think?

Reply:
I've found another case where the result doesn't match what a user might expect.

* input: `placeholder {with \} brace} in name`
* got: `placeholder  brace} in name`
* expected: `placeholder  in name`

I haven't had a look to see how nested placeholders would fair against existing tests and I think this case starts to get close to a nested braces scenario. Trying to address this risks expanding scope of the PR a bit, so am open to guidance.

Reply:
Ah, nice find... my first thought is that it's a bug, it should output what you expected.

Hopefully the fix is not too complex.

Reply:
This one still has me stumped.

The key challenge is that this line is a very convenient way of finding the closing brace.

https://github.com/caddyserver/caddy/blob/217419f6d97885daf0fb8832a407626e578523f0/replacer.go#L133

Unfortunately this means it finds closing braces even if they are escaped. Going to spend a little more time this evening and see if I can add a check to see if the closing braces was escaped.

The alternative is to re-write the `for` loop to iterate until closing braces are found. I'm trying to avoid doing this.

Reply:
I think a loop is fine; if the one that was found is preceded by a `\`, then start at that point and look ahead for the next one, until one that is not preceded by `\` is found. I bet for 99% of inputs it'll usually be the first one, so the loop won't have to iterate. When it is escaped, then there's a slight performance hit, but that's what we get for being able to escape syntactic characters...

Reply:
Thank you for outlining this approach, and specifically why it might be an acceptable trade-off. I've added this check and pushed. I also added one more test to satisfy myself I wasn't going to run into an 'off by one' scenario. New benchmarks to follow.